# Capstone Dummy Notebook

### This notebook contains the rough work of my Final Capstone Project. Here I will extract and clean data from various sources, perform various methods of exploratory analysis and draw any conclusions I find. 

### Data Requirements:

* City census data of Toronto and New York City
* Foursquare api location data


<h2><a id = "top"></a>Structure</h2>

1. <a href = "#item1"> Acquiring Census Data</a><br>

2. <a href = "#item2"> Acquiring Foursquare Data</a><br>

3. <a href = "#item3"> Creating a map of toronto </a><br>

4. <a href = "#item4"> Foursquare data for each Borough in Toronto</a><br>

5. <a href = "#item5"> Creating a Master Map for Toronto </a><br>

6. <a href = "#item6"> Analyzing Data from Master Map </a><br>

In [1]:
import pandas as pd
import numpy as np
import json # library to handle JSON files
import requests # library to handle requests
import io
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
import matplotlib.cm as cm
import matplotlib.colors as colors
import matplotlib as plt

<h2><a id = 'item1'></a> Acquiring Census Data of Toronto and New York</h2>

This is census data of the neighbourhood profiles in Toronto. I'm not using any of this data right now but I may need it later.

In [3]:
url = 'https://ckan0.cf.opendata.inter.prod-toronto.ca/download_resource/ef0239b1-832b-4d0b-a1f3-4153e53b189e?format=csv'
tor_data = pd.read_csv(url)
tor_data

_id                   Category  \
0        1  Neighbourhood Information   
1        2  Neighbourhood Information   
2        3                 Population   
3        4                 Population   
4        5                 Population   
...    ...                        ...   
2378  2379                   Mobility   
2379  2380                   Mobility   
2380  2381                   Mobility   
2381  2382                   Mobility   
2382  2383                   Mobility   

                                                 Topic  \
0                            Neighbourhood Information   
1                            Neighbourhood Information   
2                             Population and dwellings   
3                             Population and dwellings   
4                             Population and dwellings   
...                                                ...   
2378  Mobility status - Place of residence 5 years ago   
2379  Mobility status - Place of residence 5 years ago   
2380  Mobility status - Place of residence 5 years ago   
2381  Mobility status - Place of residence 5 years ago   
2382  Mobility status - Place of residence 5 years ago   

                         Data Source                    Characteristic  \
0                    City of Toronto              Neighbourhood Number   
1                    City of Toronto              TSNS2020 Designation   
2     Census Profile 98-316-X2016001                  Population, 2016   
3     Census Profile 98-316-X2016001                  Population, 2011   
4     Census Profile 98-316-X2016001       Population Change 2011-2016   
...                              ...                               ...   
2378  Census Profile 98-316-X2016001                          Migrants   
2379  Census Profile 98-316-X2016001                 Internal migrants   
2380  Census Profile 98-316-X2016001          Intraprovincial migrants   
2381  Census Profile 98-316-X2016001          Interprovincial migrants   
2382  Census Profile 98-316-X2016001                 External migrants   

     City of Toronto Agincourt North Agincourt South-Malvern West  \
0                NaN             129                          128   
1                NaN  No Designation               No Designation   
2          2,731,571          29,113                       23,757   
3          2,615,060          30,279                       21,988   
4              4.50%          -3.90%                        8.00%   
...              ...             ...                          ...   
2378         400,950           3,170                        3,145   
2379         184,120             880                          980   
2380         141,135             735                          760   
2381          42,985             135                          220   
2382         216,835           2,280                        2,170   

           Alderwood           Annex  ... Willowdale West  \
0                 20              95  ...              37   
1     No Designation  No Designation  ...  No Designation   
2             12,054          30,526  ...          16,936   
3             11,904          29,177  ...          15,004   
4              1.30%           4.60%  ...          12.90%   
...              ...             ...  ...             ...   
2378             925           6,390  ...           3,765   
2379             680           3,930  ...           1,545   
2380             615           2,630  ...           1,070   
2381              70           1,310  ...             475   
2382             245           2,460  ...           2,220   

     Willowridge-Martingrove-Richview  Woburn Woodbine Corridor  \
0                                   7     137                64   
1                      No Designation     NIA    No Designation   
2                              22,156  53,485            12,541   
3                              21,343  53,350            11,703   
4                              

This is population data of New York

In [4]:
url2 = 'https://www.baruch.cuny.edu/nycdata/population-geography/population.htm'
#https://www.baruch.cuny.edu/nycdata/population-geography/pop-demography.htm
demo = pd.read_html(url2)
demo[0]

0                                                  1   \
0  NaN                             Population & Geography   
1  NaN  New York City and Greater Metropolitan Area  P...   
2  NaN                                    Back to Section   
3  NaN  The Estimated 2018 Population of New York City...   
4  NaN                                             County   
5  NaN                                             County   
6  NaN                                                NYC   
7  NaN                                              Bronx   
8  NaN                                  Brooklyn  (Kings)   
9  NaN                              Manhattan  (New York)   
10 NaN                                             Queens   
11 NaN                          Staten Island  (Richmond)   
12 NaN                                              Total   
13 NaN                                                NaN   
14 NaN                                     New York State   
15 NaN                                             Nassau   
16 NaN                                             Putnam   
17 NaN                                           Rockland   
18 NaN                                            Suffolk   
19 NaN                                        Westchester   
20 NaN                                              Total   
21 NaN                                                NaN   
22 NaN                           New Jersey  Pennsylvania   
23 NaN                                             Bergen   
24 NaN                                              Essex   
25 NaN                                             Hudson   
26 NaN                                          Hunterdon   
27 NaN                                          Middlesex   
28 NaN                                           Monmouth   
29 NaN                                             Morris   
30 NaN                                              Ocean   
31 NaN                                            Passaic   
32 NaN                                           Somerset   
33 NaN                                             Sussex   
34 NaN                                              Union   
35 NaN                                           Pike, PA   
36 NaN                                              Total   
37 NaN                                                NaN   
38 NaN  Notes:  *Estimates are based on 2010 Census an...   
39 NaN                                                NaN   
40 NaN                                                NaN   
41 NaN  View Additional Population Data for NYC:  Popu...   
42 NaN  Source:  U.S. Census Bureau - American Communi...   
43 NaN                                                NaN   

                                                   2   \
0                              Population & Geography   
1   New York City and Greater Metropolitan Area  P...   
2                                     Back to Section   
3   The Estimated 2018 Population of New York City...   
4                                                2000   
5                                                2000   
6                                                 NaN   
7                                                1332   
8                                                2466   
9                                                1538   
10                                               2229   
11                                                444   
12                                               8009   
13                                                NaN   
14                                                NaN   
15                                               1335   
16                                                 96   
17                                                287   
18                                               1419   
19                                                924   
20                                               4061   
21  

<h2><a id = "item2"></a> Acquiring Foursquare data </h2> <a href = "#top"> back to top </a>

#### Toronto:

In [11]:
CLIENT_ID = 'NA0Q3HAIZG5G1XDGDI5ZCGQPBSNEP4SGYF5B0ULTZBDZHIOV' # your Foursquare ID
CLIENT_SECRET = 'HNVACZRJK5KNIEWX0IBRFZI0U3AG5NH1MUSWVUXBDWBFTU1J' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: NA0Q3HAIZG5G1XDGDI5ZCGQPBSNEP4SGYF5B0ULTZBDZHIOV
CLIENT_SECRET:HNVACZRJK5KNIEWX0IBRFZI0U3AG5NH1MUSWVUXBDWBFTU1J


In [12]:
# latitude and longitude of Toronto
tor_name = 'Toronto'
tor_lat = 43.6532
tor_lon = -79.3832
print('Latitude and longitude values of {} are {}, {}.'.format(tor_name, 
                                                               tor_lat, 
                                                               tor_lon))

Latitude and longitude values of Toronto are 43.6532, -79.3832.


### Explore venues in Toronto within 500 m radius


In [49]:
LIMIT = 500
radius = 10000
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    tor_lat, 
    tor_lon,
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=NA0Q3HAIZG5G1XDGDI5ZCGQPBSNEP4SGYF5B0ULTZBDZHIOV&client_secret=HNVACZRJK5KNIEWX0IBRFZI0U3AG5NH1MUSWVUXBDWBFTU1J&v=20180605&ll=43.6532,-79.3832&radius=10000&limit=500'

In [50]:
results = requests.get(url).json()

In [51]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [52]:
# results

In [53]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng', 'venue.location.postalCode']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()


c:\users\jothika pc\appdata\local\programs\python\python37-32\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name           categories        lat  \
0                  Downtown Toronto         Neighborhood  43.653232   
1  Elgin And Winter Garden Theatres              Theater  43.653394   
2                     Hailed Coffee          Coffee Shop  43.658833   
3            Nathan Phillips Square                Plaza  43.652270   
4                  Richmond Station  American Restaurant  43.651569   

         lng postalCode  
0 -79.385296        NaN  
1 -79.378507    M5B 2H1  
2 -79.383684        M5G  
3 -79.383516    M5H 2N1  
4 -79.379266        NaN

Can extract postal codes of each venue (except those that dont have one)

Can scrape this site to get boroughs of each postal code area and then merge the two tables together

In [54]:
postal_codes = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')
postal_codes = postal_codes[0]

In [55]:
postal_codes.dropna(inplace = True)

In [56]:
postal_codes

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168              Business reply mail Processing Centre  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [57]:
codes = nearby_venues['postalCode']


In [58]:
# split the postal code values to we just get the area code
for i in range(len(codes)):
    if not pd.isnull(codes[i]):
        codes[i] = codes[i].split()[0]
        

c:\users\jothika pc\appdata\local\programs\python\python37-32\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


In [59]:
len(codes)

100

In [66]:
postal_codes.set_index('Postal Code',inplace = True)

Create an array to store boroughs

#### This function takes a dataframe containing Foursquare data and will find the Borough location of each venue using its assigned postal code

In [67]:
def get_postal_borough(dataframe):
    borough = []
    codes = dataframe['postalCode']
    # split the postal code values to we just get the area code
    for i in range(len(codes)):
        if not pd.isnull(codes[i]):
            codes[i] = codes[i].split()[0]
    
    
    for i in range(len(codes)):
        if not pd.isnull(codes[i]) and codes[i] in postal_codes.index:
            borough.append(postal_codes.at[codes[i],'Borough'])
        else:
            borough.append('not assigned')
    return borough

Add this array as a column to the dataframe

In [68]:
nearby_venues['Borough'] = get_postal_borough(nearby_venues)

c:\users\jothika pc\appdata\local\programs\python\python37-32\lib\site-packages\ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys


In [69]:
nearby_venues.head(10)

name                categories        lat  \
0                      Downtown Toronto              Neighborhood  43.653232   
1      Elgin And Winter Garden Theatres                   Theater  43.653394   
2                         Hailed Coffee               Coffee Shop  43.658833   
3                Nathan Phillips Square                     Plaza  43.652270   
4                      Richmond Station       American Restaurant  43.651569   
5                     Fahrenheit Coffee               Coffee Shop  43.652384   
6                        Byblos Toronto  Mediterranean Restaurant  43.647615   
7                         Hogtown Smoke                Food Truck  43.649287   
8  St. Lawrence Market (South Building)            Farmers Market  43.648743   
9                                   Alo         French Restaurant  43.648574   

         lng postalCode           Borough  
0 -79.385296        NaN      not assigned  
1 -79.378507        M5B  Downtown Toronto  
2 -79.383684        M5G  Downtown Toronto  
3 -79.383516        M5H  Downtown Toronto  
4 -79.379266        NaN      not assigned  
5 -79.372719        M5C  Downtown Toronto  
6 -79.388381        M5V  Downtown Toronto  
7 -79.374689        M4L      East Toronto  
8 -79.371597        M5E  Downtown Toronto  
9 -79.396243        M5V  Downtown Toronto

<h2><a id = "item3"></a> Map of Toronto </h2><a href = "#top"> back to top </a>

In [70]:
# !conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library
print('Folium imported')

Folium imported


In [71]:
map_toronto = folium.Map(location=[tor_lat, tor_lon], zoom_start=10)

for lat, lng, postal, borough, name in zip(nearby_venues['lat'], nearby_venues['lng'], nearby_venues['postalCode'],nearby_venues['Borough'], nearby_venues['name']):
    label = '{},{}'.format(name,borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

<h2><a id = "item4"></a> Foursquare data for each Borough in Toronto</h2><a href = "#top"> back to top </a>

### I want to repeat this process for every borough in Toronto

In [85]:
unique_boroughs = postal_codes['Borough'].drop_duplicates()

In [86]:
# !conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

In [90]:
address = 'Etobicoke, Toronto'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates are {}, {}.'.format(latitude, longitude))

The geograpical coordinates are 43.671459150000004, -79.55249206611668.


#### This function uses the geocoders api to retrieve the coordinates of each borough in Toronto

In [91]:
def get_borough_coords():
    borough_coords = []
    for i in range(len(unique_boroughs)):
        address = unique_boroughs[i]+', Toronto'

        geolocator = Nominatim(user_agent="ny_explorer")
        location = geolocator.geocode(address)
        latitude = location.latitude
        longitude = location.longitude
#         print('The geograpical coordinates of ' + unique_boroughs[i] + ' are {}, {}.'.format(latitude, longitude))
        borough_coords.append([unique_boroughs[i],latitude,longitude])
        
    return borough_coords

In [92]:
borough_coords = get_borough_coords()

In [93]:
borough_coords

[['North York', 43.7543263, -79.44911696639593],
 ['Downtown Toronto', 43.6541737, -79.38081164513409],
 ['Etobicoke', 43.671459150000004, -79.55249206611668],
 ['Scarborough', 43.773077, -79.257774],
 ['East York', 43.699971000000005, -79.33251996261595],
 ['York', 43.67910515, -79.49118414007154],
 ['East Toronto', 43.6534817, -79.3839347],
 ['West Toronto', 43.6534817, -79.3839347],
 ['Central Toronto', 43.6534817, -79.3839347],
 ['Mississauga', 43.678523999999996, -79.62912913064454]]

The geocoders data returned the same coordinates for East Toronto, West Toronto and Central Toronto. We may decide to drop the duplicate coordinates later on and just focus on one of those three areas.  
Lets turn this into a dataframe.

In [94]:
coords_df = pd.DataFrame(borough_coords, columns = ['Borough','Lat','Lon'])
coords_df.drop_duplicates(subset = ['Lat','Lon'],inplace = True)
coords_df.reset_index(inplace = True)
coords_df.drop(columns = ['index'],inplace = True)
coords_df

Borough        Lat        Lon
0        North York  43.754326 -79.449117
1  Downtown Toronto  43.654174 -79.380812
2         Etobicoke  43.671459 -79.552492
3       Scarborough  43.773077 -79.257774
4         East York  43.699971 -79.332520
5              York  43.679105 -79.491184
6      East Toronto  43.653482 -79.383935
7       Mississauga  43.678524 -79.629129

In [105]:
def get_4square_for_boroughs_():
    i = 6

    lat = coords_df.loc[i,'Lat']
    lon = coords_df.loc[i,'Lon']
    LIMIT = 100
    radius = 500
    url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
        CLIENT_ID, 
        CLIENT_SECRET, 
        VERSION, 
        lat, 
        lon,
        radius, 
        LIMIT)
#         url # display URL
#     return url
    results = requests.get(url).json()
    results
    venues = results['response']['groups'][0]['items']
    
    df = json_normalize(venues) # flatten JSON

    # filter columns
    filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng', 'venue.location.postalCode']
    df =df.loc[:, filtered_columns]

    # filter the category for each row
    df['venue.categories'] = df.apply(get_category_type, axis=1)

    # clean columns
    df.columns = [col.split(".")[-1] for col in df.columns]
    
    # create a Borough column using the get_postal_borough funtion
    df['Borough'] = get_postal_borough(df)
    return df


In [101]:
# url = get_4square_for_boroughs()

In [102]:
# results = requests.get(url).json()

In [103]:
local_venues = get_4square_for_boroughs_()

c:\users\jothika pc\appdata\local\programs\python\python37-32\lib\site-packages\ipykernel_launcher.py:22: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
c:\users\jothika pc\appdata\local\programs\python\python37-32\lib\site-packages\ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys


In [ ]:
local_venues2 = get_4square_for_boroughs_()

In [104]:
local_venues.head(10)

name        categories        lat        lng  \
0          Downtown Toronto      Neighborhood  43.653232 -79.385296   
1    Nathan Phillips Square             Plaza  43.652270 -79.383516   
2                    Indigo         Bookstore  43.653515 -79.380696   
3              Chatime 日出茶太   Bubble Tea Shop  43.655542 -79.384684   
4  Textile Museum of Canada        Art Museum  43.654396 -79.386500   
5                 Poke Guys        Poke Place  43.654895 -79.385052   
6                   Japango  Sushi Restaurant  43.655268 -79.385165   
7                      LUSH    Cosmetics Shop  43.653557 -79.380400   
8        M Square Coffee Co       Coffee Shop  43.651218 -79.383555   
9   CF Toronto Eaton Centre     Shopping Mall  43.654540 -79.380677   

  postalCode           Borough  
0        NaN      not assigned  
1        M5H  Downtown Toronto  
2        M5B  Downtown Toronto  
3        M5G  Downtown Toronto  
4        M5G  Downtown Toronto  
5        M5G  Downtown Toronto  
6        M5G  Downtown Toronto  
7        M5B  Downtown Toronto  
8        NaN      not assigned  
9        M5B  Downtown Toronto

This is how we can merge multiple dataframes into one master frame

In [106]:
master_df = pd.DataFrame(columns = ['name','categories','lat','lng','postalCode','Borough'])

In [ ]:
frames = [local_venues,local_venues2]
master_df = pd.concat(frames)

In [ ]:
master_df

#### This function creates a dataframe for each borough with the available information retrieved from the Foursquare api, and returns a merged dataframe containing information for each borough  
Limit of 500 venues in a 10 mk radius

In [111]:
def get_4square_for_boroughs():
    i = 0
    frames = []
    while i < coords_df.shape[0]:
        lat = coords_df.loc[i,'Lat']
        lon = coords_df.loc[i,'Lon']
        LIMIT = 500
        radius = 10000
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lon,
            radius, 
            LIMIT)
    #         url # display URL
    #     return url
        results = requests.get(url).json()
        venues = results['response']['groups'][0]['items']

        df = json_normalize(venues) # flatten JSON

        # filter columns
        filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng', 'venue.location.postalCode']
        df =df.loc[:, filtered_columns]

        # filter the category for each row
        df['venue.categories'] = df.apply(get_category_type, axis=1)

        # clean columns
        df.columns = [col.split(".")[-1] for col in df.columns]

        # create a Borough column using the get_postal_borough funtion
        df['Borough'] = get_postal_borough(df)
        frames.append(df)
        i+=1
    return pd.concat(frames)

In [112]:
master_df = get_4square_for_boroughs()

c:\users\jothika pc\appdata\local\programs\python\python37-32\lib\site-packages\ipykernel_launcher.py:22: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
c:\users\jothika pc\appdata\local\programs\python\python37-32\lib\site-packages\ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys


In [113]:
 master_df.reset_index(inplace = True)

In [114]:
master_df.head()

index                       name         categories        lat        lng  \
0      0            Earl Bales Park               Park  43.753043 -79.436228   
1      1              Archers Arena  Recreation Center  43.754101 -79.467200   
2      2  Best for Bride in Toronto        Bridal Shop  43.755789 -79.437834   
3      3                       LCBO       Liquor Store  43.732717 -79.454717   
4      4        True North Climbing       Climbing Gym  43.745507 -79.474332   

  postalCode       Borough  
0        NaN  not assigned  
1        M3K    North York  
2        M3H    North York  
3        M3K    North York  
4        M3K    North York

Now we have the venue information that was available from each Borough.  There are many venues that are not assigned a Borough, we might drop these values later.  
We can also see that a large majority of these venues are coming from Downtown Toronto. 

<h2><a id = "item5"></a> Let's Create a Master Map </h2> <a href = '#top'> back to top </a>

This map will show the available venue data from all Boroughs in Toronto

In [115]:
master_map = folium.Map(location=[tor_lat, tor_lon], zoom_start=12)

for lat, lng, postal, borough, name in zip(master_df['lat'], master_df['lng'], master_df['postalCode'],master_df['Borough'], master_df['name']):
    label = '{},{}'.format(name,borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(master_map)  
    
master_map

Now we can view the venues located in each borough

<h2><a id = 'item6'></a> Analyzing Data from Master Map </h2> <a href = "#top">back to top</a>

#### Let's do some data analysis from all these venues

In [123]:
master_df['categories'].unique()

array(['Park', 'Recreation Center', 'Bridal Shop', 'Liquor Store',
       'Climbing Gym', 'Clothing Store', 'Japanese Restaurant',
       'Furniture / Home Store', 'Hotel', 'Café', 'French Restaurant',
       'Steakhouse', 'Chocolate Shop', 'Bookstore', 'Shopping Mall',
       'Airport', 'Sandwich Place', 'Ramen Restaurant',
       'Seafood Restaurant', 'Theater', 'Grocery Store', 'Massage Studio',
       'Hockey Arena', 'Warehouse Store', 'Thai Restaurant', 'Creperie',
       'Racetrack', 'Dessert Shop', 'Tea Room',
       'Middle Eastern Restaurant', 'Sushi Restaurant', 'Bakery',
       'Athletics & Sports', 'Supermarket', 'Vietnamese Restaurant',
       'Music Store', 'Flower Shop', 'Coffee Shop', 'Bagel Shop',
       'Korean Restaurant', 'Gastropub', 'Auto Dealership',
       'Bubble Tea Shop', 'Pizza Place', 'Garden', 'Deli / Bodega',
       'Gourmet Shop', 'Pharmacy', 'Field', 'Toy / Game Store',
       'Health Food Store', 'Italian Restaurant', 'Falafel Restaurant',
       'Vege

In [120]:
len(master_df['categories'].unique())

150

We have 150 unique categories. Lets clean some of this data up.  
Looking at the unique categories, we can immediately see that Toronto is dominated by ethnic restaurant chains. Lets focus on those venues for now.

In [126]:
master_df['categories']

0                   Park
1      Recreation Center
2            Bridal Shop
3           Liquor Store
4           Climbing Gym
             ...        
795         Liquor Store
796               Garden
797           Donut Shop
798        Grocery Store
799            Juice Bar
Name: categories, Length: 800, dtype: object